### ISSUE TITLE
.update() is considered to be harmful, try to reimplement with running asyncio loop in a separate thread

### ISSUE THREAD

**Issue by insolor** [insolor/async-tkinter-loop#17]
- [Update considered harmful](https://wiki.tcl-lang.org/page/Update+considered+harmful)
- [Nested Event Processing](https://tkdocs.com/tutorial/eventloop.html#nested)

**Comment by insolor** [comment]
Experimental implementation, see [separate_thread](https://github.com/insolor/async-tkinter-loop/tree/separate_thread) branch. It almost works as intended, but tests are faining, and I'm not happy that you need to switch asyncio events from a different thread, see https://github.com/insolor/async-tkinter-loop/blob/separate_thread/examples/start_stop_counter.py

**Comment by rdbende** [comment]
This is how `_tkinter.c` implements the mainloop:

https://github.com/python/cpython/blob/f63002755d3b7c0bb2e452f158769f19fcf3b850/Modules/_tkinter.c#L2694-L2724

You can see that it uses the `Tcl_DoOneEvent` function in the mainloop. Apparently this function is exposed in Tkinter, so it might be useful if you want to avoid the problems caused by `update`.

A while ago I was thinking about implementing an asynchronous mainloop in Tukaan, and this is what I came up with (transferred to Tkinter):

```py
async def _async_main_loop(root: tkinter.Tk) -> None:
    while root.is_alive:
        root.tk.dooneevent(_tkinter.DONT_WAIT)
        await asyncio.sleep(0)
```

Though using `dooneevent` at Python level is a bit slower than Tkinter's actual mainloop.

**Comment by insolor** [comment]
@rdbende thank you, I used `dooneevent` in my code. I tried idea of running the loop until `dooneevent` returns 0, but in reality `dooneevent` returns 0 most of the times, except when there are some pending events. So instead I used a nested `while root.dooneevent(_tkinter.DONT_WAIT) > 0:` loop to process all pending events in one iteration of the outer loop.

https://github.com/insolor/async-tkinter-loop/blob/4c8e5002798b72511881d63067673c8c440526b2/async_tkinter_loop.py#L16-L26
